### PR TITLE
TypeActivate MvcRouteHandler

### PR DESCRIPTION
**Pull request by ajaybhargavb**
#4292 

@rynowak

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcRouteHandler.cs [hunk 1]
```diff
@@ -10,21 +10,44 @@
...
+            : this(actionInvokerFactory, actionSelector, diagnosticSource, loggerFactory, actionContextAccessor: null)
+        {
+        }
+
+        public MvcRouteHandler(
+            IActionInvokerFactory actionInvokerFactory,
+            IActionSelector actionSelector,
+            DiagnosticSource diagnosticSource,
+            ILoggerFactory loggerFactory,
+            IActionContextAccessor actionContextAccessor)
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcRouteHandler.cs]: Let's talk about this more.

File: src/Microsoft.AspNetCore.Mvc.Core/Builder/MvcApplicationBuilderExtensions.cs [hunk 2]
```diff
@@ -89,7 +89,7 @@ public static IApplicationBuilder UseMvcWithDefaultRoute(this IApplicationBuilde
 
             var routes = new RouteBuilder(app)
             {
-                DefaultHandler = new MvcRouteHandler(),
+                DefaultHandler = app.ApplicationServices.GetRequiredService<IRouter>(),
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Builder/MvcApplicationBuilderExtensions.cs]: Register this directly as `MvcRouteHandler` - it's not the be-all-end-all of routing

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcRouteHandler.cs (hunk 1 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcRouteHandler.cs]: yes please. don't bring it back 😄

Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcRouteHandler.cs]: It's not registered by default. I was more concerned with the correct DI behavior. Is this the correct way to take an 'optional' service? or do we need to take the `IServiceProvider` instead and try to resolve it in the ctor

Comment (ajaybhargavb) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcRouteHandler.cs]: I spoke to @pranavkm. He said we can expect DI to behave correctly as it is in the spec.

**Comment by ajaybhargavb** [comment]
:up: :date:

File: src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcRouteHandler.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.Core/Internal/MvcRouteHandler.cs]: Cool :+1:

File: test/Microsoft.AspNetCore.Mvc.Core.Test/Infrastructure/MvcRouteHandlerTests.cs [hunk 3]
```diff
@@ -192,47 +204,16 @@ public class MvcRouteHandlerTests
...
-                optionsAccessor = new TestOptionsManager<MvcOptions>();
-            }
-
-            var diagnosticSource = new DiagnosticListener("Microsoft.AspNetCore");
-            if (diagnosticListener != null)
-            {
-                diagnosticSource.SubscribeWithAdapter(diagnosticListener);
-            }
+            return new MvcRouteHandler(
+                invokerFactory, actionSelector, diagnosticSource, loggerFactory, actionContextAccessor);
```
Comment (rynowak) [inline test/Microsoft.AspNetCore.Mvc.Core.Test/Infrastructure/MvcRouteHandlerTests.cs]: gross break this up

**Comment by rynowak** [comment]
:shipit: after addressing https://github.com/aspnet/Mvc/pull/4740/files/291d5d44857e871a9ba569cd9d95a4b9787b3519#r64504142

**Comment by ajaybhargavb** [comment]
I addressed that in the second commit.

**Comment by rynowak** [comment]
:shipit: again

**Pull request merged**